### DISCUSSION TITLE
Tutorial on custom block that stores post meta

### DISCUSSION THREAD

**Discussion by justintadlock** [WordPress/developer-blog-content#35] [Topic Ideas]
One of the major hurdles that plugin authors need to overcome, particularly in client work and alongside custom post types, is to learn how to create blocks that store their data as post metadata instead of within the `post_content` field.  I don't think there are a ton of resources around this in the wild, but the Developer Blog would be an ideal place to introduce this to more extenders.

Right now, this topic idea is really rough, but I want to leave it open to input on what we would want to see come out of such a tutorial.

**Comment by ryanwelcher** [comment]
@justintadlock I'd love to be part of this.

I've done a fair amount of this topic on my live stream and have a video on my YouTube channel about[ integrating with custom post types](https://www.youtube.com/watch?v=vr3wgD9IZRA). I can also provide a bunch of code for this as well.

**Reply by justintadlock** [reply]
If you want to tackle this yourself, that's cool.  Or, if you want to partner up and co-write this in some way, I'm happy to help there, too.

This is one area that I don't have any experience with yet outside reading some docs one time that mentioned it.

In my mind, I'd look at this like a specific use case for an agency/freelancer.  What's a common CPT that they would build for a client?  And, what metadata is needed for that post type?

**Reply by ryanwelcher** [reply]
> If you want to tackle this yourself, that's cool. Or, if you want to partner up and co-write this in some way, I'm happy to help there, too.

Either one works for me!

> In my mind, I'd look at this like a specific use case for an agency/freelancer. What's a common CPT that they would build for a client? And, what metadata is needed for that post type?

I can think of a couple of use-cases:

## Featured Content
in this case, a publisher site would accept external content and post it on their site but not have them as WP users.
We'd introduce a post type called Featured Content and store post meta for the author name via a Byline block or a panel in the sidebar. This would allow querying by the author's name.

## Product Testimonials
This came from a podcast I was on where they wanted to have a Testimonials block that would store the data as post meta. Then the would query for and display the product testimonials based on if there was one.

**Reply by justintadlock** [reply]
I like the testimonials idea.  I've [built a simple plugin](https://github.com/justintadlock/toot) in the past for this.  Two of the meta fields I used were:

- URL to link back to the testimonial author's site.
- Email to display the author's avatar/Gravatar.

**Comment by colorful-tones** [comment]
This is such a big barrier right now. I feel like a plugin is justified. Perhaps a Post Meta block that outputs whatever you throw at it? Sounds like a lot of work, but also telling someone who has worked with ACF to create a block all of sudden is a pretty big hurdle. 

I went down a rabbit hole to try and append custom meta to a block theme and landed on output buffering an entire template part and filtered through `the_content`. 😱  This: https://gist.github.com/colorful-tones/ddb38790de9f4ddc77f5038ba2c884f1

I love me some custom blocks and block themes, but feel like there needs to be a temporary workaround to help transition builders.

**Comment by bph** [comment]
Published: [Creating a custom block that stores post meta](https://developer.wordpress.org/news/2023/03/creating-a-custom-block-that-stores-post-meta/)
